File: app/src/main/java/com/app/cancellingretrofitrequests/java/repository/DataRepository.java
package com.app.cancellingretrofitrequests.java.repository;

import androidx.lifecycle.MutableLiveData;

import com.app.cancellingretrofitrequests.java.model.GenericDataModel;
import com.app.cancellingretrofitrequests.java.model.SearchResultModel;
import com.app.cancellingretrofitrequests.java.networking.ApiProvider;
import com.app.cancellingretrofitrequests.java.networking.ApiService;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DataRepository {
    public static ApiService
            apiService =
            new ApiProvider().createService(
                    ApiService.class);
    public static MutableLiveData<GenericDataModel<SearchResultModel>> searchResultsLiveData = new MutableLiveData<>();
    public static Call<SearchResultModel> searchResultModelCall = null;

    public static MutableLiveData<GenericDataModel<SearchResultModel>> getSearchResults(String searchQuery) {
        GenericDataModel genericDataModel =
                new GenericDataModel(
                        null,
                        null,
                        null,
                        true
                );

        searchResultsLiveData.postValue(genericDataModel);
        if (searchResultModelCall != null) {
            searchResultModelCall.cancel();
        }
        searchResultModelCall = apiService.getResult(searchQuery);
        searchResultModelCall.enqueue(new Callback<SearchResultModel>() {
            @Override
            public void onResponse(Call<SearchResultModel> call, Response<SearchResultModel> response) {
                GenericDataModel genericDataModel =
                        new GenericDataModel(
                                response.isSuccessful(),
                                response.body(),
                                response.message(),
                                false);

                searchResultsLiveData.postValue(genericDataModel);
            }

            @Override
            public void onFailure(Call<SearchResultModel> call, Throwable t) {
                if (!call.isCanceled()) {
                    GenericDataModel genericDataModel =
                            new GenericDataModel(
                                    false,
                                    null,
                                    t.getMessage(),
                                    false);

                    searchResultsLiveData.postValue(genericDataModel);
                }
            }
        });
        return searchResultsLiveData;
    }
}

File: app/src/main/java/com/app/cancellingretrofitrequests/kotlin/repository/DataRepository.kt
package com.app.cancellingretrofitrequests.kotlin.repository

import androidx.lifecycle.MutableLiveData
import com.app.cancellingretrofitrequests.kotlin.model.GenericDataModel
import com.app.cancellingretrofitrequests.kotlin.model.SearchResultModel
import com.app.cancellingretrofitrequests.kotlin.networking.ApiProvider
import com.app.cancellingretrofitrequests.kotlin.networking.ApiService

object DataRepository {
    var apiService: ApiService? = null
    var searchResultsLiveData: MutableLiveData<GenericDataModel<SearchResultModel>?> = MutableLiveData()

    init {
        apiService =
            ApiProvider.createService(
                ApiService::class.java
            )
    }

    suspend fun getSearchResults(searchQuery: String?) {
        val genericDataModelLoading =
            GenericDataModel(
                null,
                SearchResultModel(
                    null,
                    null
                ),
                null,
                true
            )
        searchResultsLiveData.postValue(genericDataModelLoading)
        val result = apiService?.getResult(searchQuery)
        val genericDataModel =
            GenericDataModel(
                result?.isSuccessful,
                result?.body(),
                result?.message(),
                false
            )
        searchResultsLiveData.postValue(genericDataModel)
    }
}